pkg/relayer/proxy/mining_supervisor.go

```diff
@@ -0,0 +1,445 @@
+package proxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	"github.com/pokt-network/poktroll/pkg/relayer/config"
+	"github.com/pokt-network/poktroll/pkg/relayer/relay_authenticator"
+	"github.com/pokt-network/poktroll/x/service/types"
+	"github.com/puzpuzpuz/xsync/v4"
+)
+
+// SessionCache stores fast-path per-session state.
+// Rewardable starts true and is only ever downgraded to false.
+type SessionCache struct {
+	EndHeight  int64
+	Rewardable atomic.Bool
+}
+
+// RelayMiningSupervisor runs delayed validation/rewardability and forwards rewardable
+// relays to the downstream "observable"/miner. It uses one bounded queue, a worker pool,
+// and an xsync-backed session cache.
+type RelayMiningSupervisor struct {
+	logger     polylog.Logger
+	downstream chan<- *types.Relay // we don't own/close this
+
+	queue chan *types.Relay // internal bounded queue
+	wg    sync.WaitGroup
+
+	ctx     context.Context
+	cancel  context.CancelFunc
+	stopped atomic.Bool
+
+	// Options
+	dropOldest          bool
+	enqueueTimeout      time.Duration
+	gaugeSampleInterval time.Duration
+	dropLogInterval     time.Duration
+
+	// Fast state
+	downstreamClosed atomic.Bool
+	lastDropLogNs    atomic.Int64 // rate-limit downstream drop logs
+
+	// Dependencies
+	relayMeter         relayer.RelayMeter
+	relayAuthenticator relayer.RelayAuthenticator
+
+	// Sessions: sessionID -> *SessionCache
+	knownSessions *xsync.Map[string, *SessionCache]
+}
+
+// NewRelayMiningSupervisor creates a new instance and starts workers + gauge sampler.
+// Panics if downstream is nil.
+func NewRelayMiningSupervisor(
+	logger polylog.Logger,
+	downstream chan<- *types.Relay,
+	cfg *config.MiningSupervisorConfig,
+	relayMeter relayer.RelayMeter,
+	relayAuthenticator relayer.RelayAuthenticator,
+) *RelayMiningSupervisor {
+	if downstream == nil {
+		// TECH_DEBT: should it panic?
+		panic("RelayMiningSupervisor: downstream channel must not be nil")
+	}
+	if cfg.QueueSize <= 0 {
+		cfg.QueueSize = 10_000
```

Comment:
If we need "magic numbers" or "magic constants" (e.g. defaults), let's move it into a local constant.